src/snapshots/numbers.md

```diff
@@ -0,0 +1,98 @@
+# META
+~~~ini
+description=Number formats
+type=expr
+~~~
+# SOURCE
+~~~roc
+(
+    0X42,
+    0x42,
+    0B01,
+    0b01,
+    0O42,
+    0o42,
+    0.1e42,
+    0.1E42,
+    0xDEADBEEF,
+    0xdeadbeef,
+    0xDeAdBeEf,
+)
+~~~
+# PROBLEMS
+**UPPERCASE BASE**
+Number base prefixes must be lowercase (0x, 0o, 0b).
+
+**UPPERCASE BASE**
+Number base prefixes must be lowercase (0x, 0o, 0b).
+
+**UPPERCASE BASE**
+Number base prefixes must be lowercase (0x, 0o, 0b).
+
+# TOKENS
+~~~zig
+OpenRound(1:1-1:2),Newline(1:1-1:1),
+Int(2:5-2:9),Comma(2:9-2:10),Newline(1:1-1:1),
+Int(3:5-3:9),Comma(3:9-3:10),Newline(1:1-1:1),
+Int(4:5-4:9),Comma(4:9-4:10),Newline(1:1-1:1),
+Int(5:5-5:9),Comma(5:9-5:10),Newline(1:1-1:1),
+Int(6:5-6:9),Comma(6:9-6:10),Newline(1:1-1:1),
+Int(7:5-7:9),Comma(7:9-7:10),Newline(1:1-1:1),
+Float(8:5-8:11),Comma(8:11-8:12),Newline(1:1-1:1),
+Float(9:5-9:11),Comma(9:11-9:12),Newline(1:1-1:1),
+Int(10:5-10:15),Comma(10:15-10:16),Newline(1:1-1:1),
+Int(11:5-11:15),Comma(11:15-11:16),Newline(1:1-1:1),
+Int(12:5-12:15),Comma(12:15-12:16),Newline(1:1-1:1),
+CloseRound(13:1-13:2),EndOfFile(13:2-13:2),
+~~~
+# PARSE
+~~~clojure
+(e-tuple @1.1-13.2
+	(e-int @2.5-2.9 (raw "0X42"))
+	(e-int @3.5-3.9 (raw "0x42"))
+	(e-int @4.5-4.9 (raw "0B01"))
+	(e-int @5.5-5.9 (raw "0b01"))
+	(e-int @6.5-6.9 (raw "0O42"))
+	(e-int @7.5-7.9 (raw "0o42"))
+	(e-frac @8.5-8.11 (raw "0.1e42"))
+	(e-frac @9.5-9.11 (raw "0.1E42"))
+	(e-int @10.5-10.15 (raw "0xDEADBEEF"))
+	(e-int @11.5-11.15 (raw "0xdeadbeef"))
+	(e-int @12.5-12.15 (raw "0xDeAdBeEf")))
+~~~
+# FORMATTED
+~~~roc
+(
+	0X42,
+	0x42,
+	0B01,
+	0b01,
+	0O42,
+	0o42,
+	0.1e42,
+	0.1E42,
+	0xDEADBEEF,
+	0xdeadbeef,
+	0xDeAdBeEf,
```

Comment:
we require `o, x, b` to be lower case, I guess the letters in exp and in hex are expected to be formatted?

Reply:
I think the main reason we error/warn for uppercase bases specifically because 0O123 looks a _lot_ like 00123. Somewhat similarly for 0B88 and 0888.

I suppose B/8 can be confused in the case of hex, so maybe there's an argument to lowercase - but I could go either way :shrug:

Reply:
i think most languages only support lowercase if they support non-base-10 literals at all.